### PR TITLE
A roadmap to TDD adoption

### PR DESCRIPTION
**Pull request by bescott1**

### CODE REVIEW

File: posts/A-roadmap-to-TDD-adoption.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+---
+authors:
+- Ben Scott
+tags:
+- Agile
+- XP
```
Comment (yhippa) [inline posts/A-roadmap-to-TDD-adoption.md]: Do we need to add "Testing" as a tag?

File: posts/A-roadmap-to-TDD-adoption.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+# An alternative approach
+
+We have to all understand that TDD is an advanced coding practice. This is something that takes time to learn and requires you to have mastered some prerequisites.
+
+First you'll need to want to write better code, this isn't something that can be forced upon you. To start your journey as a software craftsman familiarize yourself with [SOLID design principles](https://en.wikipedia.org/wiki/SOLID), and read the Clean Code book. Design is the single most important aspect of TDD. You need to design your production code, you need to design your test code and you need to design your interfaces so that they can be seamlessly be used between production and test code.
```
Comment (yhippa) [inline posts/A-roadmap-to-TDD-adoption.md]: Need some kind of link to the Clean Code book (I'm assuming by Robert C. Martin), preferably the author's page.

File: posts/A-roadmap-to-TDD-adoption.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+Once you've got those basics you can start unit testing and refactoring your code so that the tests become cleaner and more robust. Note that at this point we are not doing TDD, we are still writing tests after the code. It's important to be able to write tests after the fact first before we can really adopt TDD. There are a unique set of tools and skills in testing that need to be learned prior to be able to able to do TDD effectively. At this point you can start enforcing that any production code should be paired with its unit tests in the same commit.
+
+At this point your team should be comfortable writing tests that you can start pushing for a switch to TDD. This is where TDD exercises are effective, the team already knows how to tests their code, and how to design for testability. You can pair exercises with hand picked stories from your backlog that are especially well suited for TDD, one good example are bug fixes. Any production defect that need to be fixed should be replicated by an automated tests prior to fixing it. Not only does your team get to practice TDD but you also get the assurance that this defect won't be coming back.
```
Comment (yhippa) [inline posts/A-roadmap-to-TDD-adoption.md]: Please remove comma splices.

File: posts/A-roadmap-to-TDD-adoption.md (hunk 2 above)
Comment (bescott1) [reply posts/A-roadmap-to-TDD-adoption.md]: I didn't find one and I didn't want to link to a 3rd party like amazon

File: posts/A-roadmap-to-TDD-adoption.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+The main reason for adoption is your IT manager or tech lead has read a blog or attended a talk on how TDD solves these main problems:
+
+* Too many defects in production
+* Long QA cycles
+* New features break old features
+* Devops adoption
+
+# Traditional approach
+
+The typical adoption pattern is to start with some classroom training on how to write unit tests, and on how to mock. You follow up with some coding katas where the team has to write a common software problem using TDD (fizz-buzz, prime numbers, bowling score calculator, etc...). Once you've got a handle on those exercises you're supposed to take that back to your project and be able to develop 'real' code using TDD. This almost always fails. First, production code is much more complicated than exercises, than code that isn't written with testing in mind is ridiculously hard to do TDD with, and finally TDD isn't actually about testing and most teams focus on writing tests for test coverage and forget the most important part of the cycle. Refactoring.
```
Comment (dsharpe) [inline posts/A-roadmap-to-TDD-adoption.md]: I think "...the team has to write a common software problem using TDD..." should be something like "...the team has to solve a common software problem by writing code using TDD..."

Comment (dsharpe) [inline posts/A-roadmap-to-TDD-adoption.md]: This sentence reads confusing to me: "First, production code is much more complicated than exercises, than code that isn't written with testing in mind is ridiculously hard to do TDD with, and finally TDD isn't actually about testing and most teams focus on writing tests for test coverage and forget the most important part of the cycle. Refactoring." Maybe something like: "First, production code is much more complicated than exercises. Second, code that isn't written with testing in mind is ridiculously hard to convert to TDD. Finally, TDD isn't actually about testing and most teams focus on writing tests for test coverage and forget the most important part of the cycle."

File: posts/A-roadmap-to-TDD-adoption.md (hunk 2 above)
Comment (dsharpe) [inline posts/A-roadmap-to-TDD-adoption.md]: typo "be seamlessly be" -> "be seamlessly"

File: posts/A-roadmap-to-TDD-adoption.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+First you'll need to want to write better code, this isn't something that can be forced upon you. To start your journey as a software craftsman familiarize yourself with [SOLID design principles](https://en.wikipedia.org/wiki/SOLID), and read the Clean Code book. Design is the single most important aspect of TDD. You need to design your production code, you need to design your test code and you need to design your interfaces so that they can be seamlessly be used between production and test code.
+
+Once you've got those basics you can start unit testing and refactoring your code so that the tests become cleaner and more robust. Note that at this point we are not doing TDD, we are still writing tests after the code. It's important to be able to write tests after the fact first before we can really adopt TDD. There are a unique set of tools and skills in testing that need to be learned prior to be able to able to do TDD effectively. At this point you can start enforcing that any production code should be paired with its unit tests in the same commit.
```
Comment (dsharpe) [inline posts/A-roadmap-to-TDD-adoption.md]: "prior to be able to able to do" -> "prior to being able to do"

File: posts/A-roadmap-to-TDD-adoption.md (hunk 3 above)
Comment (dsharpe) [inline posts/A-roadmap-to-TDD-adoption.md]: Just used "at this point" in the last sentence. Maybe just "Now"?

**Pull request merged**